### PR TITLE
Offchain example

### PR DESCRIPTION
**Pull request by makoto**
## Demo

https://offchainexample2.surge.sh/name/offchainexample.eth/details

Please connect to Ropsten and you should see address, email, and description fetched from
https://offchain-resolver-example.uc.r.appspot.com/ . You can also see the same results from its subdomains such as https://offchainexample2.surge.sh/name/1.offchainexample.eth/details

## Features

- Uses ethers 5.6.2
- Display only the default cointype if resolver supports `resolves` function
- Display only the default text if resolver supports `resolves` function
- Show warning on subdomain list that it may not show the full list

## Dependencies

- https://github.com/ensdomains/ui/pull/160
- https://github.com/ensdomains/mock/pull/50
- https://github.com/ensdomains/offchain-resolver/pull/13

## NOTE

Currently `ui` and `mock` points to non official repo for testing purpose

## How to test locally

1. Setup local node

```
git clone https://github.com/ensdomains/offchain-resolver.git
cd offchain-resolver
git checkout multicoin
cd packages/contracts
yarn
npx hardhat node
```

Take notes of ENS registry contract address and the private key of the first account

2. Start gateway server
   In differnt terminal

```
cd offchain-resolver
cd packages/gateway
yarn
```

```
yarn && yarn build
yarn start --private-key 0xac0974bec39a17e36ba4a6b4d238ff944bacb478cbed5efcae784d7bf4f2ff80 --data src/offchainexample.eth.json --port 8080 --ttl 100000000
```

NOTE: Made ttl super long to prevent timeout because our local app time trip to the future when creating the seed data

3. Set up ENS app

3.1 clone the repo

```
git clone https://github.com/ensdomains/ens-app.git
cd ens-app
git checkout offchain2
yarn 
```

3.2 Create seed data

```
yarn preTest
```

3.3 start the server

```
yarn start:test
````

Access http://localhost:3000/name/1.offchainexample.eth/details and http://localhost:3000/name/offchainexample.eth/details and see it shows records

Screenshot 2022-04-11 at 16 06 57

### CODE REVIEW

**Comment by TateB** [comment]
why are empty records not showing "not set" and instead just blank?

File: cypress/integration/nameDetail.spec.js [hunk 1]
```diff
@@ -326,27 +326,32 @@ describe('Name detail view', () => {
...
         // Other Text
         .getByTestId('com.twitter-record-input')
         .clear({ force: true })
+        .wait(500)
         .type(OTHER_TEXT, { force: true })
         .wait(500)
         // Other Address
         .getByTestId('LTC-record-input', { timeout: 10000 })
         .clear({ force: true })
+        .wait(500)
```
Comment (LeonmanRolls) [inline cypress/integration/nameDetail.spec.js:354]: do we really need all of these waits?

File: cypress/integration/resolverMigration.spec.js [hunk 2]
```diff
@@ -3,7 +3,7 @@ const NAME_ROOT = Cypress.env('NAME_ROOT')
 const ENABLED_COLOUR = 'rgb(83, 132, 254)'
 const DISABLED_COLOUR = 'rgb(223, 223, 223)'
 
-describe('Migrate resolver and records', () => {
+describe.only('Migrate resolver and records', () => {
```
Comment (LeonmanRolls) [inline cypress/integration/resolverMigration.spec.js:6]: remove only

File: src/api/manager/resolvers.js [hunk 3]
```diff
@@ -10,7 +10,8 @@ import {
   getWeb3,
   isDecrypted,
   labelhash,
-  utils
+  utils,
+  foo
```
Comment (LeonmanRolls) [inline src/api/manager/resolvers.js:14]: Should we actually be calling something 'foo'?

**Comment by Arachnid** [comment]
We need to think a bit about the best way to decide when to rely on the subgraph; I don't think checking for the `resolve` method is the best option. Alternate suggestions:
 - Use the default list if the name does not have a resolver set (eg, it's a wildcard resolution).
 - Use the default list if the subgraph doesn't contain any entries for the name.

File: cypress/integration/nameDetail.spec.js (hunk 1 above)
Comment (makoto) [reply cypress/integration/nameDetail.spec.js:354]: cypress test was failing because new text was appended to the existing record without this.

**Comment by makoto** [comment]
`Use the default list if the subgraph doesn't contain any entries for the name.`

Is this always a good approach? I remember that in the ERC20 example token balance can both be on l2 or l1. You could write a resolver where a user can store data on both l1 and l2/offchain.

**Comment by makoto** [comment]
`Use the default list if the name does not have a resolver set (eg, it's a wildcard resolution).` To


1. Call  `ethers.getReslver('1.offchainexample.eth'`to see if it returns a resolver
2. Call  ENS registry contract directly (`registry.resolver("1.offchainexample.eth")`) to see if it returns a resolver
3. If #1 returns true and #2 returns 2, then `1.offchainexample.eth` is supported by wildcard

Even though this is technically possible, it's quite messy and defeats the purpose of ethers.js encapsulating the wildcard logic. Can't we think of a better way?

**Comment by makoto** [comment]
@Arachnid I made the following two changes.

- Use `provider._getResolver` to check if the given name does not have a resolver on its own. https://github.com/ensdomains/ui/pull/160/commits/4c50d22cffec3a739b0290ee15ea965b5a4af9ef
- Always append default coin lists regardless of subgraph returns coin/text keys https://github.com/ensdomains/ens-app/pull/1468/commits/463edd5b6af17978cf8f49b52f561bdd1c7bb37a

File: src/components/SingleName/ResolverAndRecords/Records.js [hunk 4]
```diff
@@ -143,24 +143,23 @@ const useGetRecords = domain => {
...
     }
   )
-
   const { loading: textRecordsLoading, data: dataTextRecords } = useQuery(
     GET_TEXT_RECORDS,
     {
       variables: {
         name: domain.name,
-        keys: resolver && resolver.texts
+        keys: _.union(resolver && resolver.texts, TEXT_PLACEHOLDER_RECORDS)
```
Comment (Arachnid) [inline src/components/SingleName/ResolverAndRecords/Records.js:158]: Will this work if `resolver` is not set? Won't it try and take the union of `undefined` and a list?

File: src/components/SingleName/SubDomains.js [hunk 5]
```diff
@@ -103,6 +105,18 @@ function SubDomains({
...
+  const { data: { wildcardResolverDomain } = {} } = useQuery(
+    WILDCARD_RESOLVER_DOMAIN,
+    {
+      variables: {
+        name: domain.name
+      }
+    }
+  )
+
+  const isWildcard =
```
Comment (Arachnid) [inline src/components/SingleName/SubDomains.js:117]: This is hard to follow. Would it be possible to rename this to something like `isInRegistry`?

**Review by Arachnid — CHANGES_REQUESTED** [review]

File: src/components/SingleName/SubDomains.js (hunk 5 above)
Comment (makoto) [reply src/components/SingleName/SubDomains.js:117]: I will get rid of this because the logic now moved as part of `wildcardResolverDomain`. However, I just realised that there is a different issue. `This domain may have other subdomains that cannot be shown as they are off chain` message can be displayed at `/name/1.offchainexample.eth/subdomains` but can NOT be displayed at `/name/offchainexample.eth/subdomains` because there is currently no way to tell that the parent domain may have wildcard subdomains.

File: src/components/SingleName/ResolverAndRecords/Records.js (hunk 4 above)
Comment (makoto) [reply src/components/SingleName/ResolverAndRecords/Records.js:158]: no, `.union()` merge an array and it also get rid of null or undefined (which I wasn't sure so it was good to double check).

```
_.union([0,1,2], [1,2,3])
(4) [0, 1, 2, 3]
_.union(null, [1,2,3])
(3) [1, 2, 3]
_.union(undefined, [1,2,3])
(3) [1, 2, 3]
```

File: src/components/SingleName/SubDomains.js (hunk 5 above)
Comment (Arachnid) [reply src/components/SingleName/SubDomains.js]: We should show that message for any domain which a) doesn't exist in the registry, or b) has a resolver that implements `resolve`. If the resolver exists and *doesn't* implement `resolve`, we know there can't be wildcard subdomains.

RE unicode support, are we changing to use Ethers in this PR?

Comment (makoto) [reply src/components/SingleName/SubDomains.js]: ` ` now indicates wildcard subdomain and its parent https://github.com/ensdomains/ui/pull/160/commits/56e9df7dcba8146559bd75e5647a5185396967a6 .

Comment (makoto) [reply src/components/SingleName/SubDomains.js]: ```We should show that message for any domain which a) doesn't exist in the registry, or b) has a resolver that implements resolve. If the resolver exists and doesn't implement resolve, we know there can't be wildcard subdomains.```

Actually I am a bit confused now. I initially checked whether the resolver has `resolve` function. I remove the check because you suggested not to rely on `resolve` function but now we do rely on?

**Comment by makoto** [comment]
NOTE: I had to disable one failing test related to name wrapper. Tracking it separately at  https://github.com/ensdomains/ens-app/issues/1500

**Review by LeonmanRolls — APPROVED** [review]

**Review by TateB — APPROVED** [review]
